### ISSUE TITLE
[FEATURE_REQUEST] memory system: let Character Cards and Character Lore be updated as the chat progresses

### ISSUE THREAD

**Issue by steve02081504** [SillyTavern/SillyTavern#2022]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

Currently a character's memory is basically context-dependent, which makes them memorize a limited number and length of conversations, and the character's setting is not permanently altered. Character X will always be character X, and no amount of effort on your part will change him one bit.
If we're letting the AI do the roleplaying dialog, why can't we have the AI summarize the dialog and update the character information?

### Describe the solution you'd like

Have the AI summarize the dialog after each round:
1. let the AI decide whether to new/update/remove Character Lore entries, and their trigger keywords
2. update the character's long term and short term goals.
3. have the AI re-summarize the content of the character descriptions and setting descriptions

### Priority

Medium (Would be very useful)

### Are you willing to test this on staging/unstable branch if this is implemented?

Yes

**Comment by Wolfsblvt** [comment]
This is already possible in ST by setting up a lore book for this and then defining "Quick Replies" that trigger a generation via STscript commands, which updates specific lore book entries.

You can find a few basic examples here: https://docs.sillytavern.app/usage/st-script/

**Comment by steve02081504** [comment]
> This is already possible in ST by setting up a lore book for this and then defining "Quick Replies" that trigger a generation via STscript commands, which updates specific lore book entries.

1. I want this feature to be automatically triggered and auto-complete in every round of dialogue instead of triggering manually.
2. Personally I think this feature is very useful, so it would be great if it could be used as a basic optional feature or plugin.

**Comment by Wolfsblvt** [comment]
Sounds like a good idea for an extension first. Like the Summary extension was in the past. Will likely depend on whether you find a willing developer who wants to tackle this.

**Comment by rosx27** [comment]
kinda had the same idea months ago, i found a sample STscript in the docs and made a few changes.

NewSummary:
```/getchatbook | /setvar key=chatLore {{char}}_chatLore |
/findentry file={{getvar::chatLore}} field=key "SummaryEntry" |
/setvar key=uid |
/getentryfield file={{getvar::chatLore}} |
/setvar key=CurrentSummary |

/setvar key=start {{lastMessageId}} |
/addvar key=start -25 |
/messages 0-{{lastMessageId}} | //use 0-{{lastMessageId}} for full, {{getvar::start}}-{{lastMessageId}} for limited
/trimtokens limit=3000 direction=end |
/setvar key=s1 |

/setvar key=summaryPrompt Pause the roleplay for a sec, You are now the narrator. Analyze the important facts and events that have happened so far in the chat between {{user}} and {{char}}. Summarize and narrate in layman’s terms specific to a 10th grader’s comprehension the facts and events in chronological order like a list of events (For example, 1. Event 1;{{newline}}2. Event 2;{{newline}}3. Event 3; and so on...). Limit the scope of your summary only to the facts and events in the chat, do not, by any means, invent anything untruthful. |

/echo Generating summary, please wait... |
/genraw lock=on instruct=off {{instructInput}}{{newline}}{{getvar::summaryPrompt}}{{newline}}{{newline}}{{instructInput}}{{newline}}These are the facts and events that happened so far:{{newline}}{{getvar::s1}}{{newline}}{{newline}}{{instructOutput}}{{newline}}The chat summary:{{newline}} |
/regex name=Thoughts-AI |
/setglobalvar key=tmp |
/run Custom.flushLore |

/flushvar CurrentSummary |
/flushvar start |
/flushvar s1 |
/flushvar summaryPrompt |
/flushvar chatLore |
/flushvar uid |
/flushvar tmp |
```

UpdateSummary:
```/getchatbook | /setvar key=chatLore {{char}}_chatLore |
/findentry file={{getvar::chatLore}} field=key "SummaryEntry" |
/setvar key=uid |
/getentryfield file={{getvar::chatLore}} |
/setvar key=CurrentSummary |

/setvar key=start {{lastMessageId}} |
/addvar key=start -25 |
/messages 0-{{lastMessageId}} | //use 0-{{lastMessageId}} for full, {{getvar::start}}-{{lastMessageId}} for limited
/trimtokens limit=3000 direction=end |
/setvar key=s1 |

/setvar key=summaryPrompt Pause the roleplay for a sec, You are now the narrator. Analyze the important facts and events that have happened so far in the chat between {{user}} and {{char}}. Summarize and narrate in layman’s terms specific to a 10th grader’s comprehension the facts and events in chronological order like a list of events (For example, 1. Event 1;{{newline}}2. Event 2;{{newline}}3. Event 3; and so on...). Limit the scope of your summary only to the facts and events in the chat, do not, by any means, invent anything untruthful. |
/echo Generating summary, please wait... |
/genraw lock=on instruct=off {{instructInput}}{{newline}}{{getvar::summaryPrompt}}{{newline}}{{newline}}{{instructInput}}{{newline}}These are the current summary: "{{getvar::CurrentSummary}}".{{newline}}Update this summary to include the latest facts and events:{{newline}}{{getvar::s1}}{{newline}}{{newline}}{{instructOutput}}{{newline}}The new chat summary (style = list of events in chronological order):{{newline}} |
/regex name=Thoughts-AI |
/setglobalvar key=tmp |
/run Custom.flushLore |

/flushvar CurrentSummary |
/flushvar start |
/flushvar s1 |
/flushvar summaryPrompt |
/flushvar chatLore |
/flushvar uid |
/flushvar tmp |
```

flushLore:
```/regex name=Thoughts-AI {{getglobalvar::tmp}} |
/setglobalvar key=cleanSummary |
/echo Retrieved! |


//setvar key=buttons "Insert" "Reject" |

//popup okButton=buttons <font color=red>Insert this new lore?/font>{{newline}}{{getglobalvar::cleanSummary}}
//if left=okButton right="Insert" rule=eq else="/echo You shall not pass \| "



/getchatbook | /setvar key=chatLore {{char}}_chatLore |
/findentry file={{getvar::chatLore}} field=key "SummaryEntry" |
/setvar key=uid |

/setentryfield file={{getvar::chatLore}} uid=uid field=content [OOC: These are the facts and events that happened so far:{{newline}}"{{getglobalvar::cleanSummary}}"{{newline}}Refer to this information to form {{char}}'s response.]{{newline}} |
/echo Summary flushed into Lorebook! |

/flushvar tmp |
/flushvar CurrentSummary |
/flushvar uid |
/flushvar cleanSummary |
```

this is a work in progress and lots of functionality that i want have not yet fully added.
